packages/xo-server-sdn-controller/src/ovsdb-client.js

```diff
@@ -0,0 +1,513 @@
+      const c = data.charAt(i)
+      if (c === '{') {
+        depth++
+      } else if (c === '}') {
+        depth--
+        if (depth === 0) {
+          const object = JSON.parse(buffer + data.substr(0, i + 1))
+          objects.push(object)
+          buffer = ''
+          data = data.substr(i + 1)
+          pos = 0
+          i = -1
+        }
+      }
+    }
+
+    buffer += data
+    return objects
+  }
+
+  // ---------------------------------------------------------------------------
+
+  async _getBridgeUuidForNetwork(networkUuid, networkName, socket) {
+    const where = [
+      [
+        'external_ids',
+        'includes',
+        ['map', [['xs-network-uuids', networkUuid]]],
+      ],
+    ]
+    const selectResult = await this._select(
+      'Bridge',
+      ['_uuid', 'name'],
+      where,
+      socket
+    )
+    if (!selectResult) {
+      return [null, null]
+    }
+
+    const bridgeUuid = selectResult._uuid[1]
+    const bridgeName = selectResult.name
+    log.debug(
+      `[${
+        this._host.name_label
+      }] Found bridge: '${bridgeName}' for network: '${networkName}'`
+    )
+
+    return [bridgeUuid, bridgeName]
+  }
+
+  async _interfaceAndPortAlreadyExist(
+    bridgeUuid,
+    bridgeName,
+    remoteAddress,
+    socket
+  ) {
+    const ports = await this._getBridgePorts(bridgeUuid, bridgeName, socket)
+    if (!ports) {
+      return
+    }
+
+    for (let i = 0; i < ports.length; ++i) {
+      const portUuid = ports[i][1]
+      const interfaces = await this._getPortInterfaces(portUuid, socket)
+      if (!interfaces) {
+        continue
+      }
+
+      let j
+      for (j = 0; j < interfaces.length; ++j) {
+        const interfaceUuid = interfaces[j][1]
+        if (
+          await this._interfaceHasRemote(interfaceUuid, remoteAddress, socket)
+        ) {
+          return true
+        }
+      }
+    }
+
+    return false
+  }
+
+  async _getBridgePorts(bridgeUuid, bridgeName, socket) {
+    const where = [['_uuid', '==', ['uuid', bridgeUuid]]]
+    const selectResult = await this._select('Bridge', ['ports'], where, socket)
+    if (!selectResult) {
+      return null
+    }
+
+    return selectResult.ports[0] === 'set'
+      ? selectResult.ports[1]
+      : [selectResult.ports]
+  }
+
+  async _getPortInterfaces(portUuid, socket) {
+    const where = [['_uuid', '==', ['uuid', portUuid]]]
+    const selectResult = await this._select(
+      'Port',
+      ['name', 'interfaces'],
+      where,
+      socket
+    )
+    if (!selectResult) {
+      return null
+    }
+
+    return selectResult.interfaces[0] === 'set'
+      ? selectResult.interfaces[1]
+      : [selectResult.interfaces]
+  }
+
+  async _interfaceHasRemote(interfaceUuid, remoteAddress, socket) {
+    const where = [['_uuid', '==', ['uuid', interfaceUuid]]]
+    const selectResult = await this._select(
+      'Interface',
+      ['name', 'options'],
+      where,
+      socket
+    )
+    if (!selectResult) {
```

Comment:
explicit the test please, try to test the expected type. apply this to all similar tests